colexec: fix an issue with builtin operators and a minor cleanup

yuzefovich opened the pull request:
Flat bytes relies on coldata.Batch.SetLength call to maintain its
invariant. We assume that it is always called before return the batch in
which Bytes vector might have been modified. This was not the case for
default builtin and substring operators, and the calls were added.
Additionally, to be safe, similar calls have been in added in projection
operators.

In a few places where we were setting the length of an internal batch to
0 and then returning it, those were replaced with returning
coldata.ZeroBatch.

Fixes: #43656.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/43989)
<!-- Reviewable:end -->

asubiotto reviewed (changes requested):
> Reviewed 14 of 14 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43989)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/and_or_projection_tmpl.go, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43989#-LydzrFO8EAl6mPcxa6Y:-LydzrFO8EAl6mPcxa6Z:bxkeo89) ([raw file](https://github.com/cockroachdb/cockroach/blob/911c22c565d9a2dc00d9d629896568c14e68e66d/pkg/sql/colexec/and_or_projection_tmpl.go#L187)):*
> ```Go
> 	// Run the right-side projection on the remaining tuples.
> 	o.rightFeedOp.batch = batch
> 	batch = o.rightProjOpChain.Next(ctx)
> ```

Would it be a possibility to copy this code above (before `origLen > 0`:
```
if origLen == 0 {
    o.rightFeed ....
    batch = o.rightProj....
    return coldata.ZeroBatch
}
```
To not have to worry about an `origLen == 0` in the rest of the code? Not sure if it can work or not.
___
*[pkg/sql/colexec/builtin_funcs.go, line 191 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43989#-LydyYTn0L6y-7jXHxWo:-LydyYToCyTUnjkonxQ_:b-yyug6d) ([raw file](https://github.com/cockroachdb/cockroach/blob/911c22c565d9a2dc00d9d629896568c14e68e66d/pkg/sql/colexec/builtin_funcs.go#L191)):*
> ```Go
> 		},
> 	)
> 	// Although we didn't change the length of the batch, it is beneficial to set
> ```

`s/beneficial/necessary`


<!-- Sent from Reviewable.io -->

yuzefovich committed:
colexec: fix an issue with builtin operators and a minor cleanup

Flat bytes relies on coldata.Batch.SetLength call to maintain its
invariant. We assume that it is always called before return the batch in
which Bytes vector might have been modified. This was not the case for
default builtin and substring operators, and the calls were added.
Additionally, to be safe, similar calls have been in added in projection
operators.

In a few places where we were setting the length of an internal batch to
0 and then returning it, those were replaced with returning
coldata.ZeroBatch. This forced us to refactor CASE and AND/OR operators
to handle cases of zero-length batches (previously, they assumed
a well-typed zero-length batch which is not always the case).

Release note: None

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43989)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/and_or_projection_tmpl.go, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43989#-LydzrFO8EAl6mPcxa6Y:-LyeHI3CAr47OP4Oifw_:b-p28k8x) ([raw file](https://github.com/cockroachdb/cockroach/blob/911c22c565d9a2dc00d9d629896568c14e68e66d/pkg/sql/colexec/and_or_projection_tmpl.go#L187)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Would it be a possibility to copy this code above (before `origLen > 0`:
```
if origLen == 0 {
    o.rightFeed ....
    batch = o.rightProj....
    return coldata.ZeroBatch
}
```
To not have to worry about an `origLen == 0` in the rest of the code? Not sure if it can work or not.
</blockquote></details>

Good point. Fixed, but we have to return `batch` (see #44017).
___
*[pkg/sql/colexec/builtin_funcs.go, line 191 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43989#-LydyYTn0L6y-7jXHxWo:-LyeGyTi3PMbWtZAHX85:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/911c22c565d9a2dc00d9d629896568c14e68e66d/pkg/sql/colexec/builtin_funcs.go#L191)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

`s/beneficial/necessary`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43989#-:-LyeKJgQ6WIwaV-1CSce:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+
